### ISSUE TITLE
JoinFunction with Collector

### ISSUE THREAD

**Issue by markus-h** [stratosphere/stratosphere#662]
The new JoinFunction does not provide a collector for the output. In some cases it could be beneficial to have such an collector as in the previous API. I propose to create an additional function (name to be discusses) that offers a Join with collector (like map <-> flatMap, reduce <-> groupReduce).
This issue came up by the implementation of a PageRank algorithm, where it is currently necessary to use an coGroup to simulate a join.

**Comment by fhueske** [comment]
I would be a bit careful to not overload the API with all possible function combinations.
Could the use-case you described be done by using the DefaultJoin and a subsequent FlatMap?

**Comment by markus-h** [comment]
Yes that can be done, but it is not very intuitive in my opinion.

**Comment by rmetzger** [comment]
:+1

Sent from my iPhone

On 07.04.2014, at 15:15, markus-h notifications@github.com wrote:

Yes that can be done, but it is not very intuitive in my opinion.

## 

Reply to this email directly or view it on
GitHubhttps://github.com/stratosphere/stratosphere/issues/662#issuecomment-39722778
.

**Comment by StephanEwen** [comment]
Can we postpone that to after the initial release?

**Comment by markus-h** [comment]
Yes. Today I had also an discussion with @fhueske whether it would be good to have the possibility to return null values for example in a cross to simulate a join with complex predicate. The way to go for now is to emit empty tuples and filter them out after the cross, since a null return ends up in an exception.
